src/board_builder/board_builder.py

```diff
@@ -0,0 +1,258 @@
+
+        for index, pose in enumerate(self.target_poses):
+            pose_values = pose.object_to_reference_matrix.values
+            pose_matrix = np.array(pose_values).reshape(4, 4)
+            corners_location = self._calculate_corners_location(pose_matrix, self.local_corners)
+            corners_dict[pose.target_id] = corners_location
+
+        return corners_dict
+
+    def build_board(self, detector_data):
+        """ Builds board using the relative matrix"""
+        detector_data = self._filter_data(detector_data)
+        if not detector_data:
+            return
+        timestamp = datetime.datetime.now()
+        corners_dict = {}
+        self.occluded_poses = []
+        self._solve_pose(detector_data, timestamp)
+        if self.target_poses:
+            for pose in self.target_poses:
+                pose_values = pose.object_to_reference_matrix.values
+                pose_matrix = np.array(pose_values).reshape(4, 4)
+                corners_location = self._calculate_corners_location(pose_matrix, self.local_corners)
+                corners_dict[pose.target_id] = corners_location
+
+            ### ID IS NOT IN FRAME ###
+            for marker_uuid in list(self._index_to_marker_uuid.values()):
+                if marker_uuid not in self._visible_markers:
+                    estimated_pose_location = PoseLocation(marker_uuid)
+                    for other_marker_pose in self.target_poses:
+                        matrix_index = self._find_matrix_input_index(other_marker_pose.target_id, marker_uuid)
+
+                        if (self._relative_pose_matrix[matrix_index[0]][matrix_index[1]] and other_marker_pose.target_id
+                                in self._visible_markers):
+                            T_AB = other_marker_pose.object_to_reference_matrix.values
+                            T_AB = np.reshape(T_AB, (4, 4))
+                            T_BC = self._relative_pose_matrix[matrix_index[0]][matrix_index[1]].get_matrix()
+                            T_AC = self._estimate_reference_to_occluded(T_AB, T_BC)
+                            estimated_pose_location.add_matrix(T_AC, str(datetime.datetime.now()))
+                    marker_pose_matrix = estimated_pose_location.get_matrix()
+                    pose = estimated_pose_location.get_pose()
+                    invisible_corners_location = self._calculate_corners_location(marker_pose_matrix,
+                                                                                  self.local_corners)
+                    corners_dict[marker_uuid] = invisible_corners_location
+                    self.occluded_poses.append(Pose(
+                        target_id=pose.target_id,
+                        object_to_reference_matrix=pose.object_to_reference_matrix,
+                        solver_timestamp_utc_iso8601=pose.solver_timestamp_utc_iso8601
+                    ))
```

Comment:
I think this code might change a little bit depending on what we discuss above (see comment on function declaration).

The point you mentioned about graph algorithms has me thinking though. We could think of the markers as nodes in a graph (which I think you already do), then have edges between all nodes that relate to some concept of strength or desirability. Then we can use something like Dijkstra or Kruskal algorithm to determine a (greedy) solution. If you haven't seen either of those before, you probably will soon in one of your classes. We can of course still consider ways to incorporate data from other transform "routes"/edges.

Reply:
I have indeed learned about Dijkstra's algorithm and weighted graphs in my free time. However, I feel that figuring out the right algorithm to determine the most desirable path could be worth an entire week on its own. For instance, if I am finding the best path from marker 1 to 3. Let's say I have only one frame where I evaluated the TMatrix from 1 to 3, but then I have 50 frames from 1 to 2, and 50 frames from 2 to 3. Which one would be more reliable, 1->3 or 1->2->3. One path is shorter but the other has more data. I feel given the time constraint, I only brainstormed the general idea, and shouldn't dive too deep while I have more important tasks to do.

Reply:
Yes, I agree that it would be time-consuming to implement different algorithms to experiment, and it is another dimension along which to experiment.

I think it is important to have *an* algorithm, and that there is an intuition/rationale behind its implementation. Others can be implemented and experimented with, but it doesn't have to be right now.